src/Microsoft.DotNet.Helix/Sdk/CheckAzurePipelinesTestResults.cs

```
@@ -39,20 +39,26 @@ private async Task CheckTestResultsAsync(HttpClient client)
         {
             foreach (var testRunId in TestRunIds)
             {
-                var data = await RetryAsync(
-                    async () =>
-                    {
-                        using (var req = new HttpRequestMessage(
-                            HttpMethod.Get,
-                            $"{CollectionUri}{TeamProject}/_apis/test/runs/{testRunId}?api-version=5.0"))
+                JObject data;
+                try
+                {
+                    data = await RetryAsync(
+                        async () =>
                         {
-                            using (var res = await client.SendAsync(req))
-                            {
-                                return await ParseResponseAsync(req, res);
-                            }
-                        }
-                    });
-                if (data != null && data["runStatistics"] is JArray runStatistics)
+                            using var req = new HttpRequestMessage(
+                                HttpMethod.Get,
+                                $"{CollectionUri}{TeamProject}/_apis/test/runs/{testRunId}?api-version=5.0");
+                            using var res = await client.SendAsync(req);
+                            return await ParseResponseAsync(req, res) ?? throw new FormatException("Helix response cannot be parsed.");
+                        });
+                }
+                catch (FormatException fe)
```

What case is this trying to catch? The only time the azure pipelines apis return non-json is if the token is bad they 302 and render the login page.

---

I didn't replicate the issue, but I started with the description of the situation at https://github.com/dotnet/arcade/issues/4220 and I tested the change on a local mock server. Do you think the logs in the ticket capture the 302 situation?

---

That issue isn't the 302 situation. We already have handling for that situation here https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs#L115.
The code has shifted a bit since that issue so the line numbers don't match up, but I guess azdev returned non-json for some reason.

Also, ParseResponseAsync logs an error in every case where it returns null, so this retry won't actually work I think. If ParseResponseAsync logs an error, the task is failed regardless of anything that happens after.

---

The way I understood -- and debugged I think -- the situation is that `ParseResponseAsync` logs an error, but swallows the exception and just return `null`. The caller then sees `null`, do not apply retry as the exception is swallowed and then -- in the original location of the problem -- fail for null dereference. The other two places I tried to fix wouldn't fail because of null, but still wouldn't apply retry logic.

---

Hm, so the environment will fail any task that will log any error during execution? Maybe I will need to schedule a quick meeting with you or some other poor soul in Monday to talk about how to run this code properly. Unit test runner doesn't seem to cut it.

---

Yea, any errors logged will cause the task to fail. Sadly the only way to really exercise this code is in the PR build because a build pipeline that has tests is required. You could extract the code and run it in a console app or LINQPad but that will require an existing azure pipelines build with tests and a PAT to access the apis.

I think it would be better to just fixup ParseResponseAsync to throw and move the handling of the 302 case into the exception handling in the base task here https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs#L71.

---

@mjanecke - thoughts on what the right pattern is to test these kinds of things in the future?  It'll be great to get these kinds of things tooled up better AND documented so there's better guidance.

---

Thanks @alexperovich , I will do the change in `ParseResponseAsync`.